### PR TITLE
[FLOC-3302] simple query api

### PR DESCRIPTION
**Pull request by avg-I**
This is a minimal implementation of the query interface.
The results paging is not supported, the sort order is fixed and only querying by a branch name is supported. The interface can be extended later on as necessary.
See FLOC-3774 for a proposed advanced query interface.

### CODE REVIEW

File: dev-requirements.txt [hunk 1]
```diff
@@ -1 +1,3 @@
 testtools==1.8.1
+python-dateutil==2.4.2
+sortedcontainers==1.4.4
```
Comment (zubron) [inline dev-requirements.txt]: These should be in `requirements.txt` rather than `dev-requirements.txt`.

File: benchmark/httpapi.py [hunk 2]
```diff
@@ -114,10 +138,12 @@ def post(self, request):
         request.setHeader(b'content-type', b'application/json')
         try:
             json = loads(request.content.read())
+            json['userdata']['branch']
```
Comment (zubron) [inline benchmark/httpapi.py]: After discussion with @jongiddy, we don't think it should be required for the user to specify a branch when submitting a new set of benchmarking results. If the branch is available for a particular set of data then it can be included in any query that includes a branch, otherwise it can be ignored or retrieved by its ID.

This came up because of the error message that follows. If `branch` is not included then the message will state that `branch` is missing but doesn't make it clear to the user that it is actually missing from `userdata`. We may not want to validate the `userdata` section.

File: benchmark/httpapi.py [hunk 3]
```diff
@@ -32,15 +37,23 @@ class ResultNotFound(Exception):
...
-
-    :ivar dict results: Stored results by their identifiers.
+    The backend that keeps the results in the memory.
     """
     def __init__(self):
-        self._results = {}
+        def get_timestamp(result):
+            return timestamp_parser.parse(result['timestamp'])
+
+        self._results = OrderedDict()
```
Comment (zubron) [inline benchmark/httpapi.py]: Does this need to be an `OrderedDict`?

File: dev-requirements.txt (hunk 1 above)
Comment (avg-I) [inline dev-requirements.txt]: Yes, that was premature. Fixed now.

File: benchmark/httpapi.py (hunk 3 above)
Comment (avg-I) [inline benchmark/httpapi.py]: No, it does not have to be `OrderedDict` now that `SortedList` is used for keeping the order.
Fixed, thanks!

File: benchmark/httpapi.py [hunk 4]
```diff
@@ -163,6 +188,55 @@ def delete(self, request, id):
...
+    @app.route("/benchmark-results", methods=['GET'])
+    def query(self, request):
+        """
+        Query the previously stored benchmarking results.
+
+        Currently this method supports filtering only by the branch name.
+        There is no support for the results paging, but a limit on the number
+        of the results is supported.
+        The order of the results is fixed at the moment and it's by
+        by the result timestamp in the descending order.
```
Comment (zubron) [inline benchmark/httpapi.py]: Remove the duplicate "by" and change "in the descending order" to just "in descending order".

File: benchmark/test/test_httpapi.py [hunk 5]
```diff
@@ -272,3 +276,139 @@ def test_delete_nonexistent(self):
...
+                       u"timestamp": datetime(2016, 1, 1, 0, 0, 7).isoformat()}
+
+    def setup_results(self):
+        """
+        Submit some results for testing various queries against them.
+        """
+        results = [
+            self.BRANCH1_RESULT1, self.BRANCH1_RESULT2, self.BRANCH2_RESULT1
+        ]
+        shuffle(results)
```
Comment (zubron) [inline benchmark/test/test_httpapi.py]: This will make the tests non-deterministic. If this change is to prove that sorting works, it would be better to insert them out of order.

File: benchmark/test/test_httpapi.py [hunk 6]
```diff
@@ -272,3 +276,139 @@ def test_delete_nonexistent(self):
...
+            d.addCallback(chained_submit, result)
+        return d
+
+    def run_query(self, ignored, filter=None, limit=None):
+        """
+        Invoke the query interface of the HTTP API.
+
+        :param dict filter: The data that the results must include.
+        :param int limit: The limit on how many results to turn.
+        :return" Deferred that fires with content of a response.
```
Comment (zubron) [inline benchmark/test/test_httpapi.py]: Change `"` to `:`

File: benchmark/test/test_httpapi.py [hunk 7]
```diff
@@ -272,3 +276,139 @@ def test_delete_nonexistent(self):
...
+        for result in results:
+            d.addCallback(chained_submit, result)
+        return d
+
+    def run_query(self, ignored, filter=None, limit=None):
+        """
+        Invoke the query interface of the HTTP API.
+
+        :param dict filter: The data that the results must include.
+        :param int limit: The limit on how many results to turn.
```
Comment (zubron) [inline benchmark/test/test_httpapi.py]: `turn` should be `return`.

File: benchmark/test/test_httpapi.py [hunk 8]
```diff
@@ -272,3 +276,139 @@ def test_delete_nonexistent(self):
...
+            query_string = ""
+        req = self.agent.request("GET", "/benchmark-results" + query_string)
+        req.addCallback(self.check_response_code, 200)
+        req.addCallback(client.readBody)
+        return req
+
+    def check_query_result(self, body, expected):
+        """
+        Check that the given response content is valid JSON
+        that contains the expect result.
```
Comment (zubron) [inline benchmark/test/test_httpapi.py]: "contains the expected result."

File: benchmark/test/test_httpapi.py [hunk 9]
```diff
@@ -272,3 +276,139 @@ def test_delete_nonexistent(self):
...
+                self.BRANCH2_RESULT1
+            ],
+        )
+        return d
+
+    def test_query_with_limit(self):
+        """
+        The latest ``limit`` results are returned if no filter is set
+        but the limit is specified and the total number of the results
+        is greater than the limit.
```
Comment (zubron) [inline benchmark/test/test_httpapi.py]: This could be simplified:
"The latest ``limit`` results are returned if no filter is set and the specified limit is less than the total number of results."

File: benchmark/test/test_httpapi.py [hunk 10]
```diff
@@ -57,7 +60,8 @@ class BenchmarkAPITests(TestCase):
     # because we test HTTP requests via an actual TCP/IP connection.
     run_tests_with = AsynchronousDeferredRunTest.make_factory(timeout=1)
 
-    RESULT = {'branch': 'branch1', 'run': 1, 'result': 1}
+    RESULT = {u"userdata": {u"branch": "master"}, u"run": 1, u"result": 1,
+              u"timestamp": datetime.now().isoformat(), }
```
Comment (zubron) [inline benchmark/test/test_httpapi.py]: Is the use of the `datetime.now()` important? If we just need a date for the timestamp field, it should probably be fixed like the later timestamp values.

File: benchmark/test/test_httpapi.py [hunk 11]
```diff
@@ -272,3 +276,139 @@ def test_delete_nonexistent(self):
...
+            ],
+        )
+        return d
+
+    def test_query_with_filter_and_limit(self):
+        """
+        The expected number of matching results is returned
+        if the total number of such results is greater than
+        the limit.  The returned results are the latest among
+        the matching results.
```
Comment (zubron) [inline benchmark/test/test_httpapi.py]: This could be simplified as well:
"The latest ``limit`` results which match the specified filter are returned if the limit is less than the total number of results."

**Comment by zubron** [comment]
Thanks, @avg-I! This looks good so far.

I was also thinking that there are a few more test cases that could probably be added. The tests for the querying currently focus on the success of the filter and limit so we should add tests to cover the error conditions that could be triggered.

Please address the comments and add the additional tests and resubmit.

**Comment by zubron** [comment]
As @avg-I is on holiday, I've picked up this issue and fixed the review comments that I made.

**Comment by zubron** [comment]
It was suggested that `eliot` should be used for logging instead of `twisted.python.log`. I've filed [FLOC-3811](https://clusterhq.atlassian.net/browse/FLOC-3811) to address this.

File: benchmark/_interfaces.py [hunk 12]
```diff
@@ -36,7 +36,9 @@ def query(filter):
         filter for the fields that are specified in the filter.
 
         :param dict filter: The filter in the JSON compatible format.
-        :return: A Deferred that fires with the results in the JSON format.
+        :param int limit: The number of the *latest* results to return.
```
Comment (sarum90) [inline benchmark/_interfaces.py]: :grey_question: Perhaps I'm just being slow today, but I don't quite follow what this means. Is it "The number of results to return. Results are sorted from most recent to least recent. Only the most recent are returned."?

That seems too wordy, but perhaps something to that extent would be more clear.

File: benchmark/httpapi.py [hunk 13]
```diff
@@ -63,14 +76,16 @@ def retrieve(self, id):
...
         """
-        matching = [
-            r for r in self._results.viewvalues()
-            if filter.viewitems() <= r.viewitems()
-        ]
+        matching = []
+        for result in reversed(self._sorted):
+            if filter.viewitems() <= result.viewitems():
+                matching.append(result)
+            if limit > 0 and len(matching) == limit:
```
Comment (sarum90) [inline benchmark/httpapi.py]: :grey_question: This behavior surprises me. I would have expected a 4XX error (maybe BadRequest? especially now that it's available) for negative limits and possibly just an empty response for limit of 0.

I don't have strong opinions on this as this appears to be somewhat of a corner case, so I'll leave this as optional, but we should potentially think about documenting this behavior in the interface definition.

File: benchmark/httpapi.py [hunk 14]
```diff
@@ -114,10 +137,11 @@ def post(self, request):
         request.setHeader(b'content-type', b'application/json')
         try:
             json = loads(request.content.read())
+            timestamp_parser.parse(json['timestamp'])
```
Comment (sarum90) [inline benchmark/httpapi.py]: :grey_question: This is sort of a design-y comment, and I'm not sure how robust the benchmark-server needs to be. So I'll leave optional for now. If implemented it should possibly even be a new PR.

It occurs to me that https://github.com/ClusterHQ/benchmark-server/pull/3/files#diff-d7efc1c0aaf2abe86cade065973bd61cR52 sort of assumes that this check has been run. It would be nice if that was codified, although that might not be worth the refactor that would be involved. Something like:

```
class StoredValue(PClass):
   timestamp = field(type=timestamp)
   json = field(type=dict)

....
  # In the interface
  def store(result):
     """
     ...
     :param StoredValue result: The result to store.
     ...
     """
...
  # And then here:
  try:
     stored_value = StoredValue(timestamp=timestamp_parse.parse(json['timestamp']),
                                                 json=json)
  except  ....
   ...
   self.backend.store(stored_value)
```

Then get_timestamp would just return value.timestamp, and would not be able to have create KeyErrors or ValueErrors, even if it gets out of sync with the code here.

File: benchmark/httpapi.py [hunk 15]
```diff
@@ -114,10 +137,11 @@ def post(self, request):
         request.setHeader(b'content-type', b'application/json')
```
Comment (sarum90) [inline benchmark/httpapi.py]: I think this now has to go after the error block (at least, I know there are HTTP server implementations that don't let you call setHeader twice for the same header, and _bad_request now also tries to set content-type.

If I'm wrong, ignore me.

File: benchmark/httpapi.py [hunk 16]
```diff
@@ -114,10 +137,11 @@ def post(self, request):
...
         try:
             json = loads(request.content.read())
+            timestamp_parser.parse(json['timestamp'])
+        except KeyError as e:
+            raise BadRequest("'{}' is missing".format(e.message))
```
Comment (sarum90) [inline benchmark/httpapi.py]: I didn't see any new tests for this new failure path, could we add a test for this? I was trying to verify my comment on line 137 and whipped up https://github.com/ClusterHQ/benchmark-server/pull/4 which surprised me when it failed.

My apologies if this is just my error and everything is fine, but that looks to me like there might be some underlying infrastructure that is not working as it should.

File: requirements.txt [hunk 17]
```diff
@@ -2,3 +2,5 @@ klein==0.2.3
 Twisted==15.5.0
 Werkzeug==0.10.4
 zope.interface==4.1.2
+python-dateutil==2.4.2
+sortedcontainers==1.4.4
```
Comment (sarum90) [inline requirements.txt]: :grey_question: Prior to this PR, the deps are sorted, perhaps we should sort these so that is maintained?

File: benchmark/httpapi.py [hunk 18]
```diff
@@ -163,6 +187,56 @@ def delete(self, request, id):
...
+
+        def got_results(results):
+            result = {"version": self.version, "results": results}
+            return dumps(result)
+
+        d.addCallback(got_results)
+        return d
+
+    @staticmethod
+    def _parse_query_args(args):
```
Comment (sarum90) [inline benchmark/httpapi.py]: :grey_question: I dunno if you've considered this (maybe it's too much overhead for this project), but we use https://python-jsonschema.readthedocs.org/en/latest/ in flocker, and it saves us from having to manually do verification like this all over the place. At this point manual verification is probably fine, but having a structured system for it is nice. If you find yourself writing too many more blocks like this we should probably switch to something like that.

**Comment by sarum90** [comment]
Looks great! I especially love the test coverage.

A couple new tests that I would like to see, otherwise I've just got some passing comments here and there. Nothing that needs action, just things to think about -- especially if this codebase grows. I think some of my optional comments would make it easier to maintain.

File: benchmark/httpapi.py (hunk 16 above)
Comment (zubron) [inline benchmark/httpapi.py]: Thanks for creating a PR for the missing tests :)

File: benchmark/httpapi.py (hunk 14 above)
Comment (zubron) [inline benchmark/httpapi.py]: You're right, I think this would be a good change to add. I won't address it as part of this PR but I'll create an issue for it.

File: benchmark/httpapi.py (hunk 18 above)
Comment (zubron) [inline benchmark/httpapi.py]: The subject of validation has already come up as it's something I would like however at the moment I don't know if adding it is the best use of the time we have. As you've mentioned though, if the number of checks starts to increase, we will probably add it.

**Comment by zubron** [comment]
Thanks for the review and the additional tests, @sarum90! :smile:

File: benchmark/httpapi.py (hunk 18 above)
Comment (sarum90) [inline benchmark/httpapi.py]: +1 to not worth time for this code. When I was reviewing I wasn't sure what level of quality we were aiming for. This seems plenty fine for its use.

File: benchmark/httpapi.py (hunk 14 above)
Comment (sarum90) [inline benchmark/httpapi.py]: Yea, actually now that I know this is just a little front-end for a internal-only test server I can go either way on that :)
